datafusion/physical-expr/src/string_expressions.rs

```diff
@@ -227,6 +229,132 @@ pub fn concat(args: &[ColumnarValue]) -> Result<ColumnarValue> {
     }
 }
 
+enum ColumnarValueRef<'a> {
+    Scalar(&'a [u8]),
+    Array(&'a StringArray),
+}
+
+impl<'a> ColumnarValueRef<'a> {
+    #[inline]
+    fn is_valid(&self, i: usize) -> bool {
+        match &self {
+            Self::Scalar(_) => true,
+            Self::Array(array) => array.is_valid(i),
+        }
+    }
+
+    #[inline]
+    fn nulls(&self) -> Option<NullBuffer> {
+        match &self {
+            Self::Scalar(_) => None,
+            Self::Array(array) => array.nulls().map(|b| b.clone()),
+        }
+    }
+}
+
+struct StringArrayBuilder {
```

Comment:
I think some comments that explained how this was different than https://docs.rs/arrow/latest/arrow/array/type.StringBuilder.html  would help. Maybe simply a note that it didn't check UTF8 again?

I wonder if we could get the same effect by adding an `unsafe` function to `StringBuilder`, like

```rust
/// Adds bytes to the in progress string, without checking for valid utf8
/// 
/// Safety: requires that bytes are valid utf8, otherwise an invalid StringArray will result
unsafe fn append_unchecked(&mut self, bytes: &[u8])

Reply:
The `Write` trait impl of StringBuilder can meet my current needs, but it is not very convenient to use, so I've defined a StringArrayBuilder. I agree with your suggestion to add an unsafe function to StringBuilder.

https://github.com/apache/arrow-rs/blob/9f36c883459405ecd9a5f4fdfa9a3317ab52302c/arrow-array/src/builder/generic_bytes_builder.rs#L231-L257
```rust
let mut builder = GenericStringBuilder::<i32>::new();

 // Write data
write!(builder, "foo").unwrap();
write!(builder, "bar").unwrap();

// Finish value
builder.append_value("baz");

// Write second value
write!(builder, "v2").unwrap();
builder.append_value("");
let array = builder.finish();

assert_eq!(array.value(0), "foobarbaz");
assert_eq!(array.value(1), "v2");
```

Reply:
Another issue is that we don't need `NullBufferBuilder` here, but `GenericByteBuilder` defaults to using `NullBufferBuilder`, which I believe introduces unnecessary overhead.